File: java-service/src/main/java/com/poc/tutorials/controller/TutorialController.java
package com.poc.tutorials.controller;

import com.poc.tutorials.model.TutorialEntity;
import com.poc.tutorials.service.TutorialService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("api/tutorials")
public class TutorialController {
    @Autowired
    private TutorialService tutorialService;

    /**
     * Adds a tutorial.
     */
    @PostMapping("/")
    public ResponseEntity<TutorialEntity> createTutorial(
            @Valid @RequestBody TutorialEntity tutorialEntity) {
        return ResponseEntity.ok(tutorialService.createTutorial(tutorialEntity));
    }

    /**
     * Delete a tutorial
     */
    @DeleteMapping(value = "/{tutorialId}")
    public ResponseEntity<HttpStatus> deleteTutorial(@PathVariable("tutorialId") UUID tutorialId) {
        return ResponseEntity.ok(tutorialService.deleteTutorialById(tutorialId) ? HttpStatus.ACCEPTED : HttpStatus.NOT_FOUND);
    }

    /**
     * Delete all tutorials
     */
    @DeleteMapping(value = "/deleteAll")
    public ResponseEntity<HttpStatus> deleteAllTutorial() {
        tutorialService.deleteAllTutorials();
        return ResponseEntity.ok(HttpStatus.ACCEPTED);
    }

    /**
     * Updates a tutorial
     */
    @PatchMapping(value = "/{tutorialId}")
    public ResponseEntity<TutorialEntity> updateTutorial(
            @PathVariable("tutorialId") UUID tutorialId,
            @Valid @RequestBody TutorialEntity tutorialEntity) {
        return ResponseEntity.ok(tutorialService.updateTutorialById(tutorialId, tutorialEntity));
    }

    /**
     * Get a tutorial by id
     */
    @GetMapping(value = "/{tutorialId}")
    public ResponseEntity<TutorialEntity> getTutorialById(
            @PathVariable("tutorialId") UUID tutorialId) {
        return tutorialService.getTutorialById(tutorialId)
                .map(t -> ResponseEntity.ok(t))
                .orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).build());
    }

    /**
     * Get all tutorials
     */
    @GetMapping(value = "/all")
    public ResponseEntity<List<TutorialEntity>> getAllTutorials(
            @RequestParam(required = false) String title) {
        System.out.println("Getting tutorials by title:  " + title);
        return ResponseEntity.ok(tutorialService.getAllTutorialsByTitle(title));
    }


    /**
     * Get tutorials published
     */

    @GetMapping(value = "/published")
    public ResponseEntity<List<TutorialEntity>> getTutorialsPublished() {
        List<TutorialEntity> tutorialsPublished = tutorialService.getTutorialsPublished();
        return tutorialsPublished.isEmpty() ? ResponseEntity.status(HttpStatus.NO_CONTENT).build()
                :
                ResponseEntity.ok(tutorialsPublished);
    }
}

File: kotlin-service/src/main/kotlin/com/example/poc/controller/TutorialController.kt
package com.example.poc.controller

import com.example.poc.model.TutorialEntity
import com.example.poc.service.TutorialService
import jakarta.validation.Valid
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*
import java.util.function.Supplier


@RestController
@RequestMapping("/api/tutorials")
@CrossOrigin("*")
class TutorialController {

    @Autowired
    private lateinit var tutorialService: TutorialService


    @GetMapping("/{id}")
    fun getById(@PathVariable id: UUID): ResponseEntity<TutorialEntity>? {
        return tutorialService.getTutorialById(id)
            .map { ResponseEntity.ok(it) }
            .orElseGet(Supplier { ResponseEntity.status(HttpStatus.NOT_FOUND).build() })
    }

    @PostMapping
    fun save(@Valid @RequestBody tutorialEntity: TutorialEntity): ResponseEntity<TutorialEntity> {
        return ResponseEntity<TutorialEntity>(tutorialService.createTutorial(tutorialEntity), HttpStatus.OK)
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable id: UUID): HttpStatus? {
        return tutorialService.getTutorialById(id)
            .map { tutorialService.deleteTutorial(it.id) }
            .map { HttpStatus.ACCEPTED }
            .orElseGet(Supplier { HttpStatus.NOT_FOUND })
    }

    @DeleteMapping("/deleteAll")
    fun deleteAll(): HttpStatus? {
        tutorialService.deleteAllTutorials()
        return HttpStatus.ACCEPTED
    }

    @PatchMapping("/{id}")
    fun updateTutorial(
        @PathVariable id: UUID,
        @Valid @RequestBody tutorialEntity: TutorialEntity
    ): ResponseEntity<TutorialEntity> {
        return ResponseEntity.ok(tutorialService.updateTutorialById(id, tutorialEntity))
    }

    @GetMapping("/all")
    fun getAllTutorialsByTitle(@RequestParam(required = false) title: String?): ResponseEntity<List<TutorialEntity>> {
        return ResponseEntity.ok( tutorialService.getAllTutorialsByTitle(title.orEmpty()) )
    }

    @GetMapping("/published")
    fun getTutorialsPublished(): ResponseEntity<List<TutorialEntity>> {
        return ResponseEntity.ok(tutorialService.getTutorialsPublished())
    }
}